### PR TITLE
Support graceful shutdown

### PR DESCRIPTION
**Pull request by pbitty**
This enables caller to signal the consumer to stop consuming messages
while allowing in-flight handlers to run to completion.

This change does not change the existing API's behaviour, except that
new errors may be returned from `Run` if `ShutDown` is called.

BEND-323

### CODE REVIEW

**Comment by pbitty** [comment]
@tonygoold I'm noting here that I think the code could use a refactor.  I think it's somewhat hard to follow as it stands, but I didn't want to block this change on a pending refactor.

File: queue_consumer.go [hunk 1]
```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
...
 }
 
+// ShutDown initiates the graceful shutdown of the consumer by stopping the ReceiveMessage loop
+// and allowing any in-flight handlers to finish processing.
+//
+// The context can be canceled during this time to abort pending operations early.  This is done co-operatively
+// and requires the provided handler func to honour the context cancelation.
+func (mf *Consumer) ShutDown() {
+	mf.shutDownLock.Lock()
+	defer mf.shutDownLock.Unlock()
```
Comment (tonygoold) [inline queue_consumer.go]: I don't think this mutex is necessary, and it points at a couple potential problems:

1. There should only be one call to `ShutDown()`. If it's being called multiple times, especially concurrently, that suggests logic errors in the caller.
2. If you close an already closed channel (or the nil channel), it will [panic](https://golang.org/ref/spec#Close).

File: example/main.go [hunk 2]
```diff
@@ -68,8 +68,16 @@ func main() {
...
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			go func() {
+				<-shutDown
+				time.AfterFunc(30*time.Second, cancel)
+				c.ShutDown()
+			}()
+
+			c.Run(ctx)
```
Comment (tonygoold) [inline example/main.go]: This strikes me as some complicated shutdown logic. You could pass `shutDown` to `Run()` and make it more analogous to the [pipeline cancellation pattern](https://blog.golang.org/pipelines), or you cancel the context without a 30 second grace period. It might just be my opinion, but I tend to think of "graceful shutdown" as meaning you can flush buffers and close open resource handles but you still have to stop immediately, with the opposite of a "graceful shutdown" being immediate process termination without cleanup.

File: queue_consumer.go [hunk 3]
```diff
@@ -121,8 +128,20 @@ func (mf *Consumer) receiveMessages(ctx context.Context, wg *sync.WaitGroup, ch
...
+// To shut down without canceling the Context, and allow in-flight messages to drain,
+// call ShutDown.
+//
+// If ShutDown has been called before Run, the returned error is ErrConsumerAlreadyShutDown.
+// If the context is canceled, the returned error is the context's error.
+// If in-flight messages drain to completion, the returned error is nil.
 func (mf *Consumer) Run(ctx context.Context) error {
+	select {
+	case <-mf.shutDown:
+		return ErrConsumerAlreadyShutDown
```
Comment (tonygoold) [inline queue_consumer.go]: I would consider a panic here, since it's significantly incorrect behaviour and probably not recoverable. A stack trace is more likely to be useful than a log message.

File: queue_consumer.go [hunk 4]
```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
...
 		select {
 		case <-ctx.Done():
-			close(jobs)
-			wg.Wait()
-			cleanupCancel()
-			cleanupWG.Wait()
 			return ctx.Err()
-		case msg := <-messages:
-			if done {
-				break
```
Comment (tonygoold) [inline queue_consumer.go]: I bet this was a bug, since it breaks from the `select`, not the enclosing `for`.

File: queue_consumer.go [hunk 5]
```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
...
 	messages := make(chan job)
-	go mf.receiveMessages(cleanupCtx, cleanupWG, messages, del)
-	var done bool
+	go mf.receiveMessages(cleanupCtx, cleanupWG, mf.shutDown, messages, del)
+	defer func() {
+		close(jobs)
+		wg.Wait()
+		cleanupCancel()
+		cleanupWG.Wait()
+	}()
```
Comment (tonygoold) [inline queue_consumer.go]: It's hard to check the correctness of code when you start passing around WaitGroups and can't easily confirm every `Add()` results in the required number of matching calls to `Done()`. I'm not sure adding even more asynchrony with a timeout is the solution, but I'd feel better seeing that than a deferred function which can potentially deadlock.

File: example/main.go (hunk 2 above)
Comment (pbitty) [reply example/main.go]: Perhaps I am overloading the term "graceful shutdown".  Pending a better term, I see two definitions:
* A "system" shutdown cleans up resources
* A "business" shutdown finishes in-progress work, then performs the "system" shutdown.
  
  This is similar to the http server's [Shutdown](https://godoc.org/net/http#Server.Shutdown), where it stops accepting new requests, but lets active requests finish.

Any ideas for a better term?  Maybe drain or stop?

File: queue_consumer.go (hunk 3 above)
Comment (pbitty) [reply queue_consumer.go]: I think we can simplify this by stopping when the channel is closed.  If we get a closed channel to begin with, we "stop" without doing any work.  This way the code paths remain the same and there are no special cases the caller has to deal with.

We usually wire our services up to listen to SIGINT and SIGTERM.  It's possible for one of those to arrive before the service has finished setting up.  We _could_ require the caller to only wire up signal handling at the end of the setup, but that seems unnecessary.

File: queue_consumer.go (hunk 4 above)
Comment (pbitty) [reply queue_consumer.go]: Ha, interesting.  It looks like it was a way to force the `<-messages` branch to be a no-op once the context was `Done`.

File: queue_consumer.go (hunk 1 above)
Comment (pbitty) [reply queue_consumer.go]: I agree on the unnecessary complexity.  My upcoming change will remove this altogether.

As far as the panic, the "lock + select" approach was meant to guard against trying to close the the already closed channel, but it may not be correct, and it's definitely not easy to ready.

File: example/main.go (hunk 2 above)
Comment (pbitty) [reply example/main.go]: Regarding the logic, I have an upcoming change to improve it.

Comment (tonygoold) [reply example/main.go]: Gotcha. It's generally better to be consistent with the standard library, so let's stick with your solution's semantics.

File: queue_consumer.go (hunk 3 above)
Comment (tonygoold) [reply queue_consumer.go]: I don't think returning without an error is the right answer. I would choose between returning an error or panicking, so the caller gets some sort of diagnostic information if they try to run an already shutdown consumer. If you don't want to panic, then I'd leave this as is.

File: queue_consumer.go (hunk 5 above)
Comment (pbitty) [reply queue_consumer.go]: Agreed.  As we discussed off Github, I will submit another PR to tackle simplifying the WaitGroup usage.

File: queue_consumer.go (hunk 3 above)
Comment (pbitty) [reply queue_consumer.go]: Ok, I would prefer to return an error, given the scenario I mentioned above.

Comment (pbitty) [reply queue_consumer.go]: I see your point. The caller at least has the ability to know why `Run` returned.

File: queue_consumer.go [hunk 6]
```diff
@@ -132,9 +137,18 @@ func (mf *Consumer) receiveMessages(ctx context.Context, wg *sync.WaitGroup, don
...
 //
 // If the context is canceled, the returned error is the context's error.
+// If the optional shutDown channel is closed before Run is called, the returned error is ErrShutdownChannelClosed.
 // If in-flight messages drain to completion after shutdown, the returned error is nil.
 func (mf *Consumer) Run(ctx context.Context, opts ...RunOption) error {
-	runOptions := resolveRunOptions(opts)
+	ro := resolveRunOptions(opts)
+	select {
+	case _, open := <-ro.shutDown:
+		if !open {
```
Comment (tonygoold) [inline queue_consumer.go]: You don't need to check the result if `ro.shutDown` is a close-only channel (the convention for "done" channels). It can simply be `case <- ro.shutDown: return ErrShutdownChannelClosed`.

Comment (pbitty) [reply queue_consumer.go]: Sounds good.  Will fix.
